pkg/virt-controller/watch/clone/clone_base.go

```diff
@@ -0,0 +1,141 @@
+	vmInformer       cache.SharedIndexInformer
+	recorder         record.EventRecorder
+
+	vmCloneQueue       workqueue.RateLimitingInterface
+	vmStatusUpdater    *status.VMStatusUpdater
+	cloneStatusUpdater *status.CloneStatusUpdater
+}
+
+func NewVmCloneController(client kubecli.KubevirtClient, vmCloneInformer, snapshotInformer, restoreInformer, vmInformer cache.SharedIndexInformer, recorder record.EventRecorder) *VMCloneController {
+	ctrl := VMCloneController{
+		client:             client,
+		vmCloneInformer:    vmCloneInformer,
+		snapshotInformer:   snapshotInformer,
+		restoreInformer:    restoreInformer,
+		vmInformer:         vmInformer,
+		recorder:           recorder,
+		vmCloneQueue:       workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "virt-controller-vmclone"),
+		vmStatusUpdater:    status.NewVMStatusUpdater(client),
+		cloneStatusUpdater: status.NewCloneStatusUpdater(client),
+	}
+
+	ctrl.vmCloneInformer.AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    ctrl.handleVMClone,
+			UpdateFunc: func(oldObj, newObj interface{}) { ctrl.handleVMClone(newObj) },
+			DeleteFunc: ctrl.handleVMClone,
+		},
+	)
+
+	return &ctrl
+}
+
+func (ctrl *VMCloneController) handleVMClone(obj interface{}) {
+	if unknown, ok := obj.(cache.DeletedFinalStateUnknown); ok && unknown.Obj != nil {
+		obj = unknown.Obj
+	}
+
+	vmClone, ok := obj.(*clonev1alpha1.VirtualMachineClone)
+	if !ok {
+		log.Log.Errorf("vm clone controller found object of unknown type")
+		return
+	}
+
+	objName, err := cache.DeletionHandlingMetaNamespaceKeyFunc(vmClone)
+	if err != nil {
+		log.Log.Errorf("vm clone controller failed to get key from object: %v, %v", err, vmClone)
+		return
+	}
+
+	log.Log.V(defaultVerbosityLevel).Infof("enqueued %q for sync", objName)
+	ctrl.vmCloneQueue.Add(objName)
+}
+
+func (ctrl *VMCloneController) Run(threadiness int, stopCh <-chan struct{}) error {
+	defer utilruntime.HandleCrash()
+	defer ctrl.vmCloneQueue.ShutDown()
+
+	log.Log.Info("Starting clone controller")
+	defer log.Log.Info("Shutting down clone controller")
+
+	if !cache.WaitForCacheSync(
+		stopCh,
+		ctrl.vmCloneInformer.HasSynced,
+		ctrl.snapshotInformer.HasSynced,
+		ctrl.restoreInformer.HasSynced,
+		ctrl.vmInformer.HasSynced,
+	) {
+		return fmt.Errorf("failed to wait for caches to sync")
+	}
+
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(ctrl.runWorker, time.Second, stopCh)
+	}
+
+	<-stopCh
+	return nil
+}
+
+func (ctrl *VMCloneController) Execute() bool {
+	key, quit := ctrl.vmCloneQueue.Get()
+	if quit {
+		return false
+	}
+	defer ctrl.vmCloneQueue.Done(key)
+
+	err, reenqueueInfo := ctrl.execute(key.(string))
+
+	if err != nil {
+		log.Log.Reason(err).Infof("reenqueuing clone %v", key)
+		ctrl.vmCloneQueue.AddRateLimited(key)
+	} else if reenqueueInfo.toReenqueue() {
+		log.Log.Infof("reenqueuing clone %v. Reason: %s", key, reenqueueInfo.reenqueueReason)
+		ctrl.vmCloneQueue.AddAfter(key, 1*time.Second)
```

Comment:
rather than requeueing every second, I think it would be better to watch VMsnapshot/VMRestore resources and when they are owned by VMClone resources, queue the owner

Reply:
:ok_hand:

Reply:
Done :ok_hand: